server/src/main/java/org/opensearch/index/translog/LocalTranslog.java

```diff
@@ -0,0 +1,232 @@
+            //
+            // For this to happen we must have already copied the translog.ckp file into translog-gen.ckp so we first check if that
+            // file exists. If not we don't even try to clean it up and wait until we fail creating it
+            assert Files.exists(nextTranslogFile) == false || Files.size(nextTranslogFile) <= TranslogHeader.headerSizeInBytes(translogUUID)
+                : "unexpected translog file: [" + nextTranslogFile + "]";
+            if (Files.exists(currentCheckpointFile) // current checkpoint is already copied
+                && Files.deleteIfExists(nextTranslogFile)) { // delete it and log a warning
+                logger.warn(
+                    "deleted previously created, but not yet committed, next generation [{}]. This can happen due to a"
+                        + " tragic exception when creating a new generation",
+                    nextTranslogFile.getFileName()
+                );
+            }
+            this.readers.addAll(recoverFromFiles(checkpoint));
+            if (readers.isEmpty()) {
+                throw new IllegalStateException("at least one reader must be recovered");
+            }
+            boolean success = false;
+            current = null;
+            try {
+                current = createWriter(
+                    checkpoint.generation + 1,
+                    getMinFileGeneration(),
+                    checkpoint.globalCheckpoint,
+                    persistedSequenceNumberConsumer
+                );
+                success = true;
+            } finally {
+                // we have to close all the recovered ones otherwise we leak file handles here
+                // for instance if we have a lot of tlog and we can't create the writer we keep on holding
+                // on to all the uncommitted tlog files if we don't close
+                if (success == false) {
+                    IOUtils.closeWhileHandlingException(readers);
+                }
+            }
+        } catch (Exception e) {
+            // close the opened translog files if we fail to create a new translog...
+            IOUtils.closeWhileHandlingException(current);
+            IOUtils.closeWhileHandlingException(readers);
+            throw e;
+        }
+    }
+
+    /**
+     * Ensures that the given location has be synced / written to the underlying storage.
+     *
+     * @return Returns <code>true</code> iff this call caused an actual sync operation otherwise <code>false</code>
+     */
+    @Override
+    public boolean ensureSynced(Location location) throws IOException {
+        try (ReleasableLock ignored = readLock.acquire()) {
+            if (location.generation == current.getGeneration()) { // if we have a new one it's already synced
+                ensureOpen();
+                return current.syncUpTo(location.translogLocation + location.size);
```

Comment:
TranslogWrite also seems to be tied heavily with local disk. Is it possible to have an abstraction (TranslogWriter) here and an LocalTranslogWriter implementation that does required job.  For RemoteLocalTranslog, the implementation could be again and extension of LocalTranslogWriter or some something similar?

Reply:
Agreed to your concerns . Currently we are doing this way as this is faster and safer way to release a working Remote Translog. Post our performance runs, if we see syncing to local disk is a significant performance hit, we will take that abstraction work , which is quite a significant work in itself.